htproxy/htproxy.go

```diff
@@ -0,0 +1,151 @@
+	"golang.org/x/net/context"
+)
+
+// HTProxy implements a dumb http proxy for the keyserver grpc interface
+type HTProxy struct {
+	Lookup func(context.Context, *proto.LookupRequest) (*proto.LookupProof, error)
+	Update func(context.Context, *proto.UpdateRequest) (*proto.LookupProof, error)
+
+	ln net.Listener
+	sr http.Server
+
+	connStateMu sync.Mutex
+	connState   map[net.Conn]http.ConnState
+
+	stopOnce sync.Once
+	stop     chan struct{}
+	waitStop sync.WaitGroup // server + all open connections
+}
+
+func (h *HTProxy) Start(ln net.Listener) {
+	h.stop = make(chan struct{})
+	h.connState = make(map[net.Conn]http.ConnState)
+	h.sr = http.Server{
+		Addr:           ln.Addr().String(),
+		Handler:        h,
+		ReadTimeout:    10 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 4096,
+		ConnState:      h.updateConnState,
+	}
+	h.ln = ln
+	h.waitStop.Add(1)
+	go h.run()
+}
+
+func (h *HTProxy) run() {
+	defer h.waitStop.Done()
+	h.sr.Serve(h.ln)
+}
+
+func (h *HTProxy) Stop() {
+	h.stopOnce.Do(func() {
+		close(h.stop)
+		h.sr.SetKeepAlivesEnabled(false)
+		h.ln.Close()
+
+		h.connStateMu.Lock()
+		for c, s := range h.connState {
+			if s == http.StateIdle {
+				c.Close()
+			}
+		}
+		h.connStateMu.Unlock()
+
+		h.waitStop.Wait()
+	})
+}
+
+func (h *HTProxy) updateConnState(c net.Conn, s http.ConnState) {
+	h.connStateMu.Lock()
+	defer h.connStateMu.Unlock()
+	h.connState[c] = s
+	switch s {
+	case http.StateNew:
+		h.waitStop.Add(1)
+	case http.StateIdle:
+		select {
+		case <-h.stop:
+			c.Close()
+		default:
+		}
+	case http.StateClosed, http.StateHijacked:
+		h.waitStop.Done()
+		delete(h.connState, c)
+	}
+}
+
+func (h *HTProxy) doLookup(b io.Reader, ctx context.Context) (*proto.LookupProof, error) {
+	lr := &proto.LookupRequest{}
+	err := jsonpb.Unmarshal(b, lr)
+	if err != nil {
+		return nil, err
+	}
+
+	pf, err := h.Lookup(ctx, lr)
+	if err != nil {
+		return nil, err
+	}
+	return pf, nil
+
+}
+
+func (h *HTProxy) doUpdate(b io.Reader, ctx context.Context) (*proto.LookupProof, error) {
+	ur := &proto.UpdateRequest{}
+	err := jsonpb.Unmarshal(b, ur)
+	if err != nil {
+		return nil, err
+	}
+
+	pf, err := h.Update(ctx, ur)
+	if err != nil {
+		return nil, err
+	}
+	return pf, nil
+
+}
+
+func (h *HTProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	path := r.URL.Path
+
+	//TODO: auth checks would go here
+
+	if r.Method != "POST" || (path != "/lookup" && path != "/update") {
+		http.Error(w, `this server only supports queries of the POST /lookup or POST /update`, http.StatusNotFound)
+		return
+	}
+	pf := &proto.LookupProof{}
+	var err error
+	ctx := context.Background()
+	if path == "/lookup" {
+		pf, err = h.doLookup(r.Body, ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	} else if path == "/update" {
+		pf, err = h.doUpdate(r.Body, ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		}
+	}
+
+	marshaler := jsonpb.Marshaler{}
```

Comment:
We can leave it as is for now. I see potentials that this HTTP proxy can be less "dump" by making the lookupProof data more friendly and favorable to the JS client.

https://github.com/yahoo/end-to-end/blob/feature/lookup-client-v2/src/javascript/crypto/e2e/coname/client.js#L188-L230 shows a list of things that can be moved from the client to this particular position at the server.